// @microsoft/teams-ai/lib/Application.js
"use strict";
/**
 * @module teams-ai
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApplicationBuilder = exports.Application = void 0;
const botbuilder_1 = require("botbuilder");
const DefaultTurnStateManager_1 = require("./DefaultTurnStateManager");
const AdaptiveCards_1 = require("./AdaptiveCards");
const MessageExtensions_1 = require("./MessageExtensions");
const AI_1 = require("./AI");
const TaskModules_1 = require("./TaskModules");
/**
 * @private
 */
const TYPING_TIMER_DELAY = 1000;
/**
 * Application class for routing and processing incoming requests.
 * @summary
 * The Application object replaces the traditional ActivityHandler that a bot would use. It supports
 * a simpler fluent style of authoring bots versus the inheritance based approach used by the
 * ActivityHandler class.
 *
 * Additionally, it has built-in support for calling into the SDK's AI system and can be used to create
 * bots that leverage Large Language Models (LLM) and other AI capabilities.
 * @template TState Optional. Type of the turn state. This allows for strongly typed access to the turn state.
 */
class Application {
    _options;
    _routes = [];
    _invokeRoutes = [];
    _adaptiveCards;
    _messageExtensions;
    _taskModules;
    _ai;
    _beforeTurn = [];
    _afterTurn = [];
    _typingTimer;
    /**
     * Creates a new Application instance.
     * @param {ApplicationOptions<TState>} options Optional. Options used to configure the application.
     */
    constructor(options) {
        this._options = Object.assign({
            removeRecipientMention: true,
            startTypingTimer: true,
            longRunningMessages: false
        }, options);
        // Create default turn state manager if needed
        if (!this._options.turnStateManager) {
            this._options.turnStateManager = new DefaultTurnStateManager_1.DefaultTurnStateManager();
        }
        // Create AI component if configured with a planner
        if (this._options.ai) {
            this._ai = new AI_1.AI(this._options.ai);
        }
        this._adaptiveCards = new AdaptiveCards_1.AdaptiveCards(this);
        this._messageExtensions = new MessageExtensions_1.MessageExtensions(this);
        this._taskModules = new TaskModules_1.TaskModules(this);
        // Validate long running messages configuration
        if (this._options.longRunningMessages && (!this._options.adapter || !this._options.botAppId)) {
            throw new Error(`The Application.longRunningMessages property is unavailable because no adapter or botAppId was configured.`);
        }
    }
    /**
     * Fluent interface for accessing Adaptive Card specific features.
     * @returns {AdaptiveCards<TState>} The AdaptiveCards instance.
     */
    get adaptiveCards() {
        return this._adaptiveCards;
    }
    /**
     * Fluent interface for accessing AI specific features.
     * @summary
     * This property is only available if the Application was configured with `ai` options. An
     * exception will be thrown if you attempt to access it otherwise.
     * @returns {AI<TState>} The AI instance.
     */
    get ai() {
        if (!this._ai) {
            throw new Error(`The Application.ai property is unavailable because no AI options were configured.`);
        }
        return this._ai;
    }
    /**
     * Fluent interface for accessing Message Extensions' specific features.
     * @returns {MessageExtensions<TState>} The MessageExtensions instance.
     */
    get messageExtensions() {
        return this._messageExtensions;
    }
    /**
     * The application's configured options.
     * @returns {ApplicationOptions<TState>} The application's configured options.
     */
    get options() {
        return this._options;
    }
    /**
     * Fluent interface for accessing Task Module specific features.
     * @returns {TaskModules<TState>} The TaskModules instance.
     */
    get taskModules() {
        return this._taskModules;
    }
    /**
     * Adds a new route to the application.
     * @summary
     * Developers won't typically need to call this method directly as it's used internally by all
     * of the fluent interfaces to register routes for their specific activity types.
     *
     * Routes will be matched in the order they're added to the application. The first selector to
     * return `true` when an activity is received will have its handler called.
     *
     * Invoke-based activities receive special treatment and are matched separately as they typically
     * have shorter execution timeouts.
     * @param {RouteSelector} selector Function thats used to select a route. The function should return true to trigger the route.
     * @param {RouteHandler<TState>} handler Function to call when the route is triggered.
     * @param {boolean} isInvokeRoute Optional. Boolean indicating if the RouteSelector is for an activity that uses "invoke" which require special handling. Defaults to `false`.
     * @returns {this} The application instance for chaining purposes.
     */
    addRoute(selector, handler, isInvokeRoute = false) {
        if (isInvokeRoute) {
            this._invokeRoutes.push({ selector, handler });
        }
        else {
            this._routes.push({ selector, handler });
        }
        return this;
    }
    /**
     * Handles incoming activities of a given type.
     * @param {string | RegExp | RouteSelector | string[] | RegExp[] | RouteSelector[]} type Name of the activity type to match or a regular expression to match against the incoming activity type. An array of type names or expression can also be passed in.
     * @param {(context: TurnContext, state: TState) => Promise<void>} handler Function to call when the route is triggered.
     * @param {TurnContext} handler.context The context object for the turn.
     * @param {TState} handler.state The state object for the turn.
     * @returns {this} The application instance for chaining purposes.
     */
    activity(type, handler) {
        (Array.isArray(type) ? type : [type]).forEach((t) => {
            const selector = createActivitySelector(t);
            this.addRoute(selector, handler);
        });
        return this;
    }
    /**
     * Handles conversation update events.
     * @param {ConversationUpdateEvents | ConversationUpdateEvents[]} event Name of the conversation update event(s) to handle.
     * @param {(context: TurnContext, state: TState) => Promise<void>} handler Function to call when the route is triggered.
     * @param {TurnContext} handler.context The context object for the turn.
     * @param {TState} handler.state The state object for the turn.
     * @returns {this} The application instance for chaining purposes.
     */
    conversationUpdate(event, handler) {
        if (typeof handler !== 'function') {
            throw new Error(`ConversationUpdate 'handler' for ${event} is ${typeof handler}. Type of 'handler' must be a function.`);
        }
        (Array.isArray(event) ? event : [event]).forEach((e) => {
            const selector = createConversationUpdateSelector(e);
            this.addRoute(selector, handler);
        });
        return this;
    }
    async continueConversationAsync(context, logic) {
        if (!this._options.adapter) {
            throw new Error(`You must configure the Application with an 'adapter' before calling Application.continueConversationAsync()`);
        }
        if (!this._options.botAppId) {
            console.warn(`Calling Application.continueConversationAsync() without a configured 'botAppId'. In production environments a 'botAppId' is required.`);
        }
        // Identify conversation reference
        let reference;
        if (typeof context.activity == 'object') {
            reference = botbuilder_1.TurnContext.getConversationReference(context.activity);
        }
        else if (typeof context.type == 'string') {
            reference = botbuilder_1.TurnContext.getConversationReference(context);
        }
        else {
            reference = context;
        }
        await this._options.adapter.continueConversationAsync(this._options.botAppId ?? '', reference, logic);
    }
    /**
     * Handles incoming messages with a given keyword.
     * @summary
     * This method provides a simple way to have a bot respond anytime a user sends your bot a
     * message with a specific word or phrase.
     *
     * For example, you can easily clear the current conversation anytime a user sends "/reset":
     *
     * ```JavaScript
     * bot.message('/reset', async (context, state) => {
     *     await state.conversation.delete();
     *     await context.sendActivity(`I have reset your state.`);
     * });
     * ```
     * @param {string | RegExp | RouteSelector | string[] | RegExp[] | RouteSelector[]} keyword Substring of text or a regular expression to match against the text of an incoming message. An array of keywords or expression can also be passed in.
     * @param {(context: TurnContext, state: TState) => Promise<void>} handler Function to call when the route is triggered.
     * @returns {this} The application instance for chaining purposes.
     */
    message(keyword, handler) {
        (Array.isArray(keyword) ? keyword : [keyword]).forEach((k) => {
            const selector = createMessageSelector(k);
            this.addRoute(selector, handler);
        });
        return this;
    }
    /**
     * Handles message reaction events.
     * @param {MessageReactionEvents | MessageReactionEvents[]} event Name of the message reaction event(s) to handle.
     * @param {(context: TurnContext, state: TState) => Promise<void>} handler Function to call when the route is triggered.
     * @returns {this} The application instance for chaining purposes.
     */
    messageReactions(event, handler) {
        (Array.isArray(event) ? event : [event]).forEach((e) => {
            const selector = createMessageReactionSelector(e);
            this.addRoute(selector, handler);
        });
        return this;
    }
    /**
     * Dispatches an incoming activity to a handler registered with the application.
     * @summary
     * This method should be called from your bot's "turn handler" (its primary message handler)
     *
     * ```JavaScript
     * server.post('/api/messages', async (req, res) => {
     *    await adapter.processActivity(req, res, async (context) => {
     *      await bot.run(context);
     *   });
     * });
     * ```
     * @param {TurnContext} turnContext Context class for the current turn of conversation with the user.
     * @returns {Promise<boolean>} True if the activity was successfully dispatched to a handler. False if no matching handlers could be found.
     */
    async run(turnContext) {
        return await this.startLongRunningCall(turnContext, async (context) => {
            // Start typing indicator timer
            this.startTypingTimer(context);
            try {
                // Remove @mentions
                if (this._options.removeRecipientMention && context.activity.type == botbuilder_1.ActivityTypes.Message) {
                    context.activity.text = botbuilder_1.TurnContext.removeRecipientMention(context.activity);
                }
                // Load turn state
                const { storage, turnStateManager } = this._options;
                const state = await turnStateManager.loadState(storage, context);
                // Call beforeTurn event handlers
                if (!(await this.callEventHandlers(context, state, this._beforeTurn))) {
                    // Save turn state
                    // - This lets the bot keep track of why it ended the previous turn. It also
                    //   allows the dialog system to be used before the AI system is called.
                    await turnStateManager.saveState(storage, context, state);
                    return false;
                }
                // Run any RouteSelectors in this._invokeRoutes first if the incoming Teams activity.type is "Invoke".
                // Invoke Activities from Teams need to be responded to in less than 5 seconds.
                if (context.activity.type === botbuilder_1.ActivityTypes.Invoke) {
                    for (let i = 0; i < this._invokeRoutes.length; i++) {
                        // TODO: fix security/detect-object-injection
                        // eslint-disable-next-line security/detect-object-injection
                        const route = this._invokeRoutes[i];
                        if (await route.selector(context)) {
                            // Execute route handler
                            await route.handler(context, state);
                            // Call afterTurn event handlers
                            if (await this.callEventHandlers(context, state, this._afterTurn)) {
                                // Save turn state
                                await turnStateManager.saveState(storage, context, state);
                            }
                            // End dispatch
                            return true;
                        }
                    }
                }
                // All other ActivityTypes and any unhandled Invokes are run through the remaining routes.
                for (let i = 0; i < this._routes.length; i++) {
                    // TODO:
                    // eslint-disable-next-line security/detect-object-injection
                    const route = this._routes[i];
                    if (await route.selector(context)) {
                        // Execute route handler
                        await route.handler(context, state);
                        // Call afterTurn event handlers
                        if (await this.callEventHandlers(context, state, this._afterTurn)) {
                            // Save turn state
                            await turnStateManager.saveState(storage, context, state);
                        }
                        // End dispatch
                        return true;
                    }
                }
                // Call AI module if configured
                if (this._ai && context.activity.type == botbuilder_1.ActivityTypes.Message && context.activity.text) {
                    // Begin a new chain of AI calls
                    await this._ai.chain(context, state);
                    // Call afterTurn event handlers
                    if (await this.callEventHandlers(context, state, this._afterTurn)) {
                        // Save turn state
                        await turnStateManager.saveState(storage, context, state);
                    }
                    // End dispatch
                    return true;
                }
                // activity wasn't handled
                return false;
            }
            finally {
                this.stopTypingTimer();
            }
        });
    }
    async sendProactiveActivity(context, activityOrText, speak, inputHint) {
        let response;
        await this.continueConversationAsync(context, async (ctx) => {
            response = await ctx.sendActivity(activityOrText, speak, inputHint);
        });
        return response;
    }
    /**
     * Manually start a timer to periodically send "typing" activities.
     * @summary
     * The timer waits 1000ms to send its initial "typing" activity and then send an additional
     * "typing" activity every 1000ms. The timer will automatically end once an outgoing activity
     * has been sent. If the timer is already running or the current activity, is not a "message"
     * the call is ignored.
     * @param {TurnContext} context The context for the current turn with the user.
     */
    startTypingTimer(context) {
        if (context.activity.type == botbuilder_1.ActivityTypes.Message && !this._typingTimer) {
            // Listen for outgoing activities
            context.onSendActivities((context, activities, next) => {
                // Listen for any messages to be sent from the bot
                if (timerRunning) {
                    for (let i = 0; i < activities.length; i++) {
                        // TODO:
                        // eslint-disable-next-line security/detect-object-injection
                        if (activities[i].type == botbuilder_1.ActivityTypes.Message) {
                            // Stop the timer
                            this.stopTypingTimer();
                            timerRunning = false;
                            break;
                        }
                    }
                }
                return next();
            });
            let timerRunning = true;
            const onTimeout = async () => {
                try {
                    // Send typing activity
                    await context.sendActivity({ type: botbuilder_1.ActivityTypes.Typing });
                }
                catch (err) {
                    // Seeing a random proxy violation error from the context object. This is because
                    // we're in the middle of sending an activity on a background thread when the turn ends.
                    // The context object throws when we try to update "this.responded = true". We can just
                    // eat the error but lets make sure our states cleaned up a bit.
                    this._typingTimer = undefined;
                    timerRunning = false;
                }
                // Restart timer
                if (timerRunning) {
                    this._typingTimer = setTimeout(onTimeout, TYPING_TIMER_DELAY);
                }
            };
            this._typingTimer = setTimeout(onTimeout, TYPING_TIMER_DELAY);
        }
    }
    /**
     * Manually stop the typing timer.
     * @summary
     * If the timer isn't running nothing happens.
     */
    stopTypingTimer() {
        if (this._typingTimer) {
            clearTimeout(this._typingTimer);
            this._typingTimer = undefined;
        }
    }
    /**
     * Registers a turn event handler.
     * @summary
     * Turn events let you do something before or after a turn is run. Returning false from
     * `beforeTurn` lets you prevent the turn from running and returning false from `afterTurn`
     * lets you prevent the bots state from being saved.
     *
     * Returning false from `beforeTurn` does result in the bots state being saved which lets you
     * track the reason why the turn was not processed. It also means you can use `beforeTurn` as
     * a way to call into the dialog system. For example, you could use the OAuthPrompt to sign the
     * user in before allowing the AI system to run.
     * @param {TurnEvents | TurnEvents[]} event - Name of the turn event to handle.
     * @param {(context: TurnContext, state: TState) => Promise<boolean>} handler - Function to call when the event is triggered.
     * @returns {this} The application instance for chaining purposes.
     */
    turn(event, handler) {
        (Array.isArray(event) ? event : [event]).forEach((e) => {
            switch (event) {
                case 'beforeTurn':
                default:
                    this._beforeTurn.push(handler);
                    break;
                case 'afterTurn':
                    this._afterTurn.push(handler);
                    break;
            }
        });
        return this;
    }
    /**
     * Calls the given event handlers with the given context and state.
     * @param {TurnContext} context - The context for the current turn with the user.
     * @param {TState} state - The current state of the conversation.
     * @param {ApplicationEventHandler<TState>[]} handlers - The event handlers to call.
     * @returns {Promise<boolean>} A Promise that resolves to a boolean indicating whether the event handlers completed successfully.
     * @private
     */
    async callEventHandlers(context, state, handlers) {
        for (let i = 0; i < handlers.length; i++) {
            // TODO:
            // eslint-disable-next-line security/detect-object-injection
            const continueExecution = await handlers[i](context, state);
            if (!continueExecution) {
                return false;
            }
        }
        // Continue execution
        return true;
    }
    /**
     * Calls the given handler with the given context, either directly or by continuing the conversation
     * if the message is a long-running message.
     * @param {TurnContext} context - The context for the current turn with the user.
     * @param {(context: TurnContext) => Promise<boolean>} handler - The handler function to call.
     * @returns {Promise<boolean>} A Promise that resolves to a boolean indicating whether the handler completed successfully.
     * @private
     */
    startLongRunningCall(context, handler) {
        /**
         * If the message is a long-running message, continue the conversation
         * and call the handler with the new context.
         */
        if (context.activity.type == botbuilder_1.ActivityTypes.Message && this._options.longRunningMessages) {
            return new Promise((resolve, reject) => {
                this.continueConversationAsync(context, async (ctx) => {
                    try {
                        // Copy original activity to new context
                        for (const key in context.activity) {
                            // eslint-disable-next-line security/detect-object-injection
                            ctx.activity[key] = context.activity[key];
                        }
                        // Call handler
                        const result = await handler(ctx);
                        resolve(result);
                    }
                    catch (err) {
                        reject(err);
                    }
                });
            });
        }
        else {
            // Call handler directly
            return handler(context);
        }
    }
}
exports.Application = Application;
/**
 * A builder class for simplifying the creation of an Application instance.
 * @template TState Optional. Type of the turn state. This allows for strongly typed access to the turn state.
 */
class ApplicationBuilder {
    _options = {};
    /**
     * Configures the application to use long running messages.
     * Default state for longRunningMessages is false
     * @param {BotAdapter} adapter The adapter to use for routing incoming requests.
     * @param {string} botAppId The Microsoft App ID for the bot.
     * @returns {this} The ApplicationBuilder instance.
     */
    withLongRunningMessages(adapter, botAppId) {
        if (!botAppId) {
            throw new Error(`The Application.longRunningMessages property is unavailable because botAppId cannot be null or undefined.`);
        }
        this._options.longRunningMessages = true;
        this._options.adapter = adapter;
        this._options.botAppId = botAppId;
        return this;
    }
    /**
     * Configures the storage system to use for storing the bot's state.
     * @param {Storage} storage The storage system to use.
     * @returns {this} The ApplicationBuilder instance.
     */
    withStorage(storage) {
        this._options.storage = storage;
        return this;
    }
    /**
     * Configures the AI system to use for processing incoming messages.
     * @param {AIOptions<TState>} aiOptions The options for the AI system.
     * @returns {this} The ApplicationBuilder instance.
     */
    withAIOptions(aiOptions) {
        this._options.ai = aiOptions;
        return this;
    }
    /**
     * Configures the turn state manager to use for managing the bot's turn state.
     * @param {TurnStateManager<TState>} turnStateManager The turn state manager to use.
     * @returns {this} The ApplicationBuilder instance.
     */
    withTurnStateManager(turnStateManager) {
        this._options.turnStateManager = turnStateManager;
        return this;
    }
    /**
     * Configures the processing of Adaptive Card requests.
     * @param {AdaptiveCardsOptions} adaptiveCardOptions The options for the Adaptive Cards.
     * @returns {this} The ApplicationBuilder instance.
     */
    withAdaptiveCardOptions(adaptiveCardOptions) {
        this._options.adaptiveCards = adaptiveCardOptions;
        return this;
    }
    /**
     * Configures the processing of Task Module requests.
     * @param {TaskModulesOptions} taskModuleOptions The options for the Task Modules.
     * @returns {this} The ApplicationBuilder instance.
     */
    withTaskModuleOptions(taskModuleOptions) {
        this._options.taskModules = taskModuleOptions;
        return this;
    }
    /**
     * Configures the removing of mentions of the bot's name from incoming messages.
     * Default state for removeRecipientMention is true
     * @param {boolean} removeRecipientMention The boolean for removing reciepient mentions.
     * @returns {this} The ApplicationBuilder instance.
     */
    setRemoveRecipientMention(removeRecipientMention) {
        this._options.removeRecipientMention = removeRecipientMention;
        return this;
    }
    /**
     * Configures the typing timer when messages are received.
     * Default state for startTypingTimer is true
     * @param {boolean} startTypingTimer The boolean for starting the typing timer.
     * @returns {this} The ApplicationBuilder instance.
     */
    setStartTypingTimer(startTypingTimer) {
        this._options.startTypingTimer = startTypingTimer;
        return this;
    }
    /**
     * Builds and returns a new Application instance.
     * @returns {Application<TState>} The Application instance.
     */
    build() {
        return new Application(this._options);
    }
}
exports.ApplicationBuilder = ApplicationBuilder;
/**
 * @param {string | RegExp | RouteSelector} type The type of activity to match. Can be a string, RegExp, or RouteSelector function.
 * @returns {RouteSelector} A RouteSelector function that matches the given activity type.
 * @private
 */
function createActivitySelector(type) {
    if (typeof type == 'function') {
        // Return the passed in selector function
        return type;
    }
    else if (type instanceof RegExp) {
        // Return a function that matches the activities type using a RegExp
        return (context) => {
            return Promise.resolve(context?.activity?.type ? type.test(context.activity.type) : false);
        };
    }
    else {
        // Return a function that attempts to match type name
        const typeName = type.toString().toLocaleLowerCase();
        return (context) => {
            return Promise.resolve(context?.activity?.type ? context.activity.type.toLocaleLowerCase() === typeName : false);
        };
    }
}
/**
 * Creates a route selector function that matches a conversation update event.
 * @param {ConversationUpdateEvents} event The conversation update event to match against.
 * @returns {RouteSelector} A route selector function that returns true if the activity is a conversation update event and matches the specified event type.
 * @private
 */
function createConversationUpdateSelector(event) {
    switch (event) {
        case 'channelCreated':
        case 'channelDeleted':
        case 'channelRenamed':
        case 'channelRestored':
            /**
             * @param {TurnContext} context The context object for the current turn of conversation.
             * @returns {Promise<boolean>} A Promise that resolves to a boolean indicating whether the activity is a conversation update event related to channels.
             */
            return (context) => {
                return Promise.resolve(context?.activity?.channelId === 'msteams' &&
                    context?.activity?.type == botbuilder_1.ActivityTypes.ConversationUpdate &&
                    context?.activity?.channelData?.eventType == event &&
                    context?.activity?.channelData?.channel &&
                    context.activity.channelData?.team);
            };
        case 'membersAdded':
            /**
             * @param {TurnContext} context The context object for the current turn of conversation.
             * @returns {Promise<boolean>} A Promise that resolves to a boolean indicating whether the activity is a conversation update event with members added.
             */
            return (context) => {
                return Promise.resolve(context?.activity?.type == botbuilder_1.ActivityTypes.ConversationUpdate &&
                    Array.isArray(context?.activity?.membersAdded) &&
                    context.activity.membersAdded.length > 0);
            };
        case 'membersRemoved':
            /**
             * @param {TurnContext} context The context object for the current turn of conversation.
             * @returns {Promise<boolean>} A Promise that resolves to a boolean indicating whether the activity is a conversation update event with members removed.
             */
            return (context) => {
                return Promise.resolve(context?.activity?.type == botbuilder_1.ActivityTypes.ConversationUpdate &&
                    Array.isArray(context?.activity?.membersRemoved) &&
                    context.activity.membersRemoved.length > 0);
            };
        case 'teamRenamed':
        case 'teamDeleted':
        case 'teamHardDeleted':
        case 'teamArchived':
        case 'teamUnarchived':
        case 'teamRestored':
            /**
             * @param {TurnContext} context The context object for the current turn of conversation.
             * @returns {Promise<boolean>} A Promise that resolves to a boolean indicating whether the activity is a conversation update event related to teams.
             */
            return (context) => {
                return Promise.resolve(context?.activity?.channelId === 'msteams' &&
                    context?.activity?.type == botbuilder_1.ActivityTypes.ConversationUpdate &&
                    context?.activity?.channelData?.eventType == event &&
                    context?.activity?.channelData?.team);
            };
        default:
            /**
             * @param {TurnContext} context The context object for the current turn of conversation.
             * @returns {Promise<boolean>} A Promise that resolves to a boolean indicating whether the activity is a conversation update event with the specified event type.
             */
            return (context) => {
                return Promise.resolve(context?.activity?.type == botbuilder_1.ActivityTypes.ConversationUpdate &&
                    context?.activity?.channelData?.eventType == event);
            };
    }
}
/**
 * Creates a route selector function that matches a message based on a keyword.
 * @param {string | RegExp | RouteSelector} keyword The keyword to match against the message text. Can be a string, regular expression, or a custom selector function.
 * @returns {RouteSelector} A route selector function that returns true if the message text matches the keyword.
 * @private
 */
function createMessageSelector(keyword) {
    if (typeof keyword == 'function') {
        // Return the passed in selector function
        return keyword;
    }
    else if (keyword instanceof RegExp) {
        // Return a function that matches a messages text using a RegExp
        return (context) => {
            if (context?.activity?.type === botbuilder_1.ActivityTypes.Message && context.activity.text) {
                return Promise.resolve(keyword.test(context.activity.text));
            }
            else {
                return Promise.resolve(false);
            }
        };
    }
    else {
        // Return a function that attempts to match a messages text using a substring
        const k = keyword.toString().toLocaleLowerCase();
        return (context) => {
            if (context?.activity?.type === botbuilder_1.ActivityTypes.Message && context.activity.text) {
                return Promise.resolve(context.activity.text.toLocaleLowerCase().indexOf(k) >= 0);
            }
            else {
                return Promise.resolve(false);
            }
        };
    }
}
/**
 * @param {MessageReactionEvents} event The type of message reaction event to create a selector for.
 * @returns {RouteSelector} A selector function that matches the specified message reaction event.
 * @private
 */
function createMessageReactionSelector(event) {
    switch (event) {
        case 'reactionsAdded':
        default:
            /**
             * @param {TurnContext} context The context object for the current turn of the conversation.
             * @returns {Promise<boolean>} A promise that resolves to true if the context object represents a message reaction event with reactions added, or false otherwise.
             */
            return (context) => {
                return Promise.resolve(context?.activity?.type == botbuilder_1.ActivityTypes.MessageReaction &&
                    Array.isArray(context?.activity?.reactionsAdded) &&
                    context.activity.reactionsAdded.length > 0);
            };
        case 'reactionsRemoved':
            /**
             * @param {TurnContext} context The context object for the current turn of the conversation.
             * @returns {Promise<boolean>} A promise that resolves to true if the context object represents a message reaction event with reactions removed, or false otherwise.
             */
            return (context) => {
                return Promise.resolve(context?.activity?.type == botbuilder_1.ActivityTypes.MessageReaction &&
                    Array.isArray(context?.activity?.reactionsRemoved) &&
                    context.activity.reactionsRemoved.length > 0);
            };
    }
}
